Refactor: Supabase Javascript v2 Migration

SiravitPhokeed opened the pull request:
**Features**
- Migration to Supabase Javascript Library v2
- Migration to Supabase Next.js Auth Helpers and Supabase React Auth Helpers
  Resolves #89
- Supabase RLS support
- `.env.local.example` file guides on creating the local `.env.local` file
- Elective Period text transition
- Migration to a central Supabase type file generated by the Supabase CLI, slightly modified to allow for join tables support
- Elective Period now reveals associated Subject Periods on hover, not on click
  - This is still on touch on mobile, the Close Button is still present
- Class pages now utilizes ISR
- Remove non-functional components to avoid confusion
  (The code is commented out and can be uncommented once functionality is implemented)
  - Remove News Filters in News
  - Remove Edit Info in Account
- In grouped imports, rename `Modules` to `External libraries`

**Fixes**
- Access token is never refreshed
  Fixes #88
- Citizen ID validator in Welcome does not add the 12th digit to the sum, causing an incorrect expected checksum and ultimately the incorrect validation result
- Elective Period in Schedule Lookup translation fails to show
- Rename `useTeacherOption` to `useTeacherOptions` to show that there are, in fact, many teacher options
- Rename the misspelled `getCurrentAcedemicYear` to `getCurrentAcademicYear`
- Remove references to other locale type solutions than `LangCode`
- Use `T[]` notation instead of `Array<T>`

**Notes to QA**
This PR affects every page. I request you test every, single, page.

Jimmy-Tempest commented:
A few questions before reviewing the PR: Would this break the implementation of PR #91, how should we go about fixing the merge conflicts, and which PR should be merged first?

SiravitPhokeed commented:
> Would this break the implementation of PR #91?

Most likely not. On most files I just updated the type annotation and pass in the Supabase client from `getServerSideProps` to allow for RLS support.

> How should we go about fixing the merge conflicts?

I would be surprised if there is any. I have already updated this branch with `main` and everything seems to work fine. I guess #91 might produce some but, again, unlikely. As you are the owner of that PR and thus knows the most about that PR, you can go ahead and fix those conflicts, if they happen.

> Which PR should be merged first?

I don’t think it matters, but I reckon we get the big one out of the way first.

Jimmy-Tempest commented:
My concern is that there could be a supabase v2 API change that would break the new types added by PR #91 so I think merging smaller PR could be better in terms of minimizing potential conflict.

Jimmy-Tempest commented on .env.local.template:
```diff
@@ -0,0 +1,3 @@
+NEXT_PUBLIC_SUPABASE_URL=
```
This file might become obsolete with the addition of `global.d.ts` which can define type of env

SiravitPhokeed replied:
The point of this file is to tell new developers that you do need a `.env.local` file and get them started.

Jimmy-Tempest commented on src/pages/account/index.tsx:
```diff
@@ -332,7 +334,8 @@ const AccountDetails: NextPage<{ user: Student | Teacher }> = ({ user }) => {
           toggleShowLogOut={toggleShowLogOut}
         />
-        <EditInfoSection user={user} />
+        {/* TODO: Edit Info functionality */}
```
Make sure this is created as an issue

Jimmy-Tempest commented on src/pages/subject/[subjectID].tsx:
```diff
@@ -754,7 +748,7 @@ const SubjectDetails: NextPage<{
           setDeleteConnection={setDeleteConnection}
         />
-        <PeriodLogsSection
+        {/* <PeriodLogsSection
```
Should this be a TODO or should it be remove all together

SiravitPhokeed replied:
It looks like they don’t want us to keep these features, but might as well just commented it out just in case they want them again.

Jimmy-Tempest reviewed (approved):
LGTM!!! However, there is something I want to request like documentation of how to work on the project after this as in what steps you need to take to run and build the project considering you will need to generate type now if I understand correctly.

SiravitPhokeed commented:
There will be conflicts, the merge order just affects who fixes the issue.

Jimmy-Tempest commented:
Coin flip?

SiravitPhokeed commented:
> Coin flip?

Live share.

Jimmy-Tempest commented:
> > Coin flip?
> 
> Live share.

There's no fun in that

SiravitPhokeed commented:
> what steps you need to take to run and build the project considering you will need to generate type now

The central database types file was originally generated with the Supabase CLI but has to be modified to support joining tables. It’s because of this modification that the types file can’t just be regenerated every build time. If the database has been altered, it’s probably easier to just manually modify the file.

Temp9699 commented:
Please see #101 
I hope it's just something wrong from my side but It's causing me trouble to continue doing QA for this PR

Temp9699 reviewed (approved):
lgtm
